src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.MetadataDb.cs

```diff
@@ -165,23 +146,52 @@ public void Dispose()
                 Length = 0;
             }
 
-            internal void TrimExcess()
+            /// <summary>
+            /// If using array pools, trim excess if necessary.
+            /// If not using array pools, release the temporary array pool and alloc.
+            /// </summary>
+            /// <param name="useArrayPools"></param>
+            internal void CompleteAllocations(bool useArrayPools)
             {
-                // There's a chance that the size we have is the size we'd get for this
-                // amount of usage (particularly if Enlarge ever got called); and there's
-                // the small copy-cost associated with trimming anyways. "Is half-empty" is
-                // just a rough metric for "is trimming worth it?".
-                if (Length <= _data.Length / 2)
+                if (useArrayPools)
                 {
-                    byte[] newRent = ArrayPool<byte>.Shared.Rent(Length);
-                    byte[] returnBuf = newRent;
-
-                    if (newRent.Length < _data.Length)
+                    // There's a chance that the size we have is the size we'd get for this
+                    // amount of usage (particularly if Enlarge ever got called); and there's
+                    // the small copy-cost associated with trimming anyways. "Is half-empty" is
+                    // just a rough metric for "is trimming worth it?".
+                    if (Length <= _data.Length / 2)
                     {
-                        Buffer.BlockCopy(_data, 0, newRent, 0, Length);
-                        returnBuf = _data;
-                        _data = newRent;
+                        byte[] newRent = ArrayPool<byte>.Shared.Rent(Length);
+                        byte[] returnBuf = newRent;
+
+                        if (newRent.Length < _data.Length)
+                        {
+                            Buffer.BlockCopy(_data, 0, newRent, 0, Length);
+                            returnBuf = _data;
+                            _data = newRent;
+                        }
+
+                        // The data in this rented buffer only conveys the positions and
+                        // lengths of tokens in a document, but no content; so it does not
+                        // need to be cleared.
+                        ArrayPool<byte>.Shared.Return(returnBuf);
                     }
+                }
+                else
+                {
+                    // This is faster and allocates less than creating a new MetadataDb.
```

Comment:
Can you add a bit more detail, specifying that the new MetadataDb would have been created when cloning a JsonElement with a parent (as done previously in the object and JsonElement converters)?

Reply:
This code has changed; I don't believe extra doc helps here any longer

Reply:
Added a table that describes all cases